### ISSUE TITLE
Segfault on startup with 0.11.1

### ISSUE THREAD

**Issue by DeMarko** [Shopify/ruby-lsp#1063]
### Operating System

Mac OS 13.6 (22G120)

### Ruby version

3.2.2

### Project has a bundle

- [X] Has bundle

### Ruby version manager being used

rbenv

### Description

This started yesterday for me after restarting my VSCode. I noticed that on start I would see two notices of the Ruby LSP failing when I started up my editor. 
![Screenshot 2023-10-04 at 5 21 04 PM](https://github.com/Shopify/vscode-ruby-lsp/assets/173373/a63cadab-dfd9-4545-8554-40a98d5ae9f8)

<details><summary>Details</summary>

Ruby LSP> Skipping custom bundle setup since /Users/danneljurado/dev/kickstarter/.ruby-lsp/Gemfile.lock already exists and is up to date
Ruby LSP> Running bundle install for the custom bundle. This may take a while...
Resolving dependencies...
The Gemfile's dependencies are satisfied
Starting Ruby LSP...
/Users/danneljurado/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/ruby-lsp-0.11.1/lib/ruby_indexer/lib/ruby_indexer/index.rb:164: [BUG] Segmentation fault at 0x000000011dddd000
ruby 3.2.2 (2023-03-30 revision e51014f9c0) +YJIT [x86_64-darwin22]
<output omitted>
[IMPORTANT]
Don't forget to include the Crash Report log file under
DiagnosticReports directory in bug reports.

[Error - 5:04:37 PM] Client Ruby LSP: connection to server is erroring. Shutting down server.
[Error - 5:04:37 PM] Sending document notification textDocument/didOpen failed
Error: write EPIPE
	at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16)
[Error - 5:04:37 PM] Request textDocument/diagnostic failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Document pull failed for text document file:///Users/danneljurado/dev/kickstarter/app/mailers/user_mailer/presenters.rb
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Connection to server got closed. Server will not be restarted.
[Error - 5:04:37 PM] Request textDocument/documentSymbol failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Request textDocument/codeAction failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Request textDocument/documentLink failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Request textDocument/inlayHint failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Request textDocument/semanticTokens/range failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Request textDocument/foldingRange failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Request textDocument/codeLens failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Request textDocument/semanticTokens/full failed.
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Stopping server failed
  Message: Cannot call write after a stream was destroyed
  Code: -32099 
[Error - 5:04:37 PM] Stopping server failed
  Message: Cannot call write after a stream was destroyed
  Code: -32099 

</details> 

I clicked on the output and it appears to be a segfault of some sort? I've attached the DiagnosticReport file (I had to zip it because GitHub)
[ruby_2023-10-04-164704_Dixie-Flatline.diag.zip](https://github.com/Shopify/vscode-ruby-lsp/files/12809142/ruby_2023-10-04-164704_Dixie-Flatline.diag.zip)

I am on a 2.3 GHz Quad-Core Intel Core i7 Macbook Pro (I think it's an early 2020). It was working just fine a couple of days ago as I was able to use the Rubocop quickfix feature while editing code.

Lemme know if I can get you any other details to help diagnose this!

**Comment by RSO** [comment]
Not sure if this is still helpful, but I'm seeing the same behavior, ruby-lsp crashes instantly:
- MacOS 14.0
- Ruby 3.2.2
- ruby-lsp 0.11.1
- Version manager: rbenv
- With bundle

[Output from VSCode](https://pastebin.com/mHNefEVi)

**Comment by andyw8** [comment]
It looks like the problem happens while indexing the code. Could you please try add the following script to your project as `index.rb`, then run `BUNDLE_GEMFILE=.ruby-lsp/Gemfile bundle exec ruby index.rb`

```ruby
require "ruby-lsp"
require "ruby_lsp/internal"

module RubyIndexer
  index = Index.new

  RubyIndexer.configuration.indexables.each do |indexable|
    puts "indexing: #{indexable.full_path}"
    content = File.read(indexable.full_path)
    visitor = IndexVisitor.new(index, YARP.parse(content), indexable.full_path)
    visitor.run
  end
end
```

That should help determine if it's being crashing for a particular file.

**Comment by RSO** [comment]
> It looks like the problem happens while indexing the code. Could you please try add the following script to your project as `index.rb`, then run `BUNDLE_GEMFILE=.ruby-lsp/Gemfile bundle exec ruby index.rb`
> 
> ```ruby
> require "ruby-lsp"
> require "ruby_lsp/internal"
> 
> module RubyIndexer
>   index = Index.new
> 
>   RubyIndexer.configuration.indexables.each do |indexable|
>     puts "indexing: #{indexable.full_path}"
>     content = File.read(indexable.full_path)
>     visitor = IndexVisitor.new(index, YARP.parse(content), indexable.full_path)
>     visitor.run
>   end
> end
> ```
> 
> That should help determine if it's crashing for a particular file.

Running this does not result in any failure, unfortunately.

**Comment by andyw8** [comment]
@DeMarko oh, very interesting, I will look into that gem.

**Comment by DeMarko** [comment]
> Thanks for the report. Does it crash as soon as you start working, or only after some period?

Sorry, I missed this earlier. Mine crashes immediately when I start working, the status reported in VSCode is that it's trying to index files before the crash

**Comment by a-lavis** [comment]
At first when I ran the script earlier, it worked fine like @RSO said.  But I tried again just now, and it failed.  Similar to my earlier post, it seems non-deterministic on my machine; sometimes it fails, sometimes it doesn't.

I ran `BUNDLE_GEMFILE=.ruby-lsp/Gemfile bundle exec ruby index.rb &> ~/logs/lsp.txt` and after two attempts it failed and outputted:
```
[1]    21724 abort      BUNDLE_GEMFILE=.ruby-lsp/Gemfile bundle exec ruby index.rb &> ~/logs/lsp.txt
```
Here's that `~/logs/lsp.txt` file: [lsp.txt](https://github.com/Shopify/ruby-lsp/files/12820318/lsp.txt)

Here's the crash report: [ruby-2023-10-05-113201.ips.zip](https://github.com/Shopify/ruby-lsp/files/12820332/ruby-2023-10-05-113201.ips.zip)

**Comment by andyw8** [comment]
Interesting, I can't reproduce it for `launchy`.

It looks like `graphql-pro` is a commercial gem so I can't access the code, would you mind checking if there are any occurrences of `frozen_string_literals` ?

**Comment by frederikspang** [comment]
Oh my god.
We had 3 files in our project, that had the (well known.... :) ) magic comment:
```
# frozen_string_literals: true
```

Spot one mistake there! Fixing the spelling worked for us.

**Comment by dalizard** [comment]
I am experiencing the same (segmentation fault) problem and after patching the local copy of a few gems using `frozen_string_literals` I haven't seen a crash 😅 Hopefully this fixes the problem. Thank you @andyw8!

**Comment by steveclarke** [comment]
I'm having a segmentation fault error in one of my codebases as well. After doing some triage on my Gemfile, I narrowed it down to the "guard" gem. Something in that gem or one of its dependencies causes the segfault every time. Remove the gem and "Ruby LSP is ready".

**Comment by RSO** [comment]
Taking a step back here, do we actually think it's ok for the parser to crash out like this when there's a typo in a comment in the file? 🤔 I was thinking that if a file does not have the correct `frozen_string_literal` comment, the comment would be just that: a comment. Or is there something else that I'm missing here?

**Comment by andyw8** [comment]
https://github.com/ruby/prism/issues/1629 has been fixed. So once there is a new Prism release, and we update the dependency in Ruby LSP, this will be resolved.

**Comment by andyw8** [comment]
We're working on updating, should be available soon. https://github.com/Shopify/ruby-lsp/pull/1098

**Comment by andyw8** [comment]
Note: The `guard` gem also has a dependency on `lumberjack` so this will likely impact a lot of projects.

**Comment by vinistock** [comment]
Just released [v0.12](https://github.com/Shopify/ruby-lsp/releases/tag/v0.12.0) which should address this issue. Sorry for the trouble, folks.

**Comment by DeMarko** [comment]
Just updated! Thank you for fixing this! 😄
